### PR TITLE
PLT-1604: Add some benchmarks for summing using 'data lists'

### PR DESCRIPTION
**Pull request by michaelpj**
Matthias says that Aiken encodes _all_ structured data as `Data`. I was very surprised that this is performant at all, so I thought I would write a benchmark to test.

I wrote a variant of our existing list benchmarks that just uses explicit operations on `Data` to mimic a list of integers.

Indeed, "data lists" seem to be significantly worse (3-4x) than both Scott-encoding and builtin lists. That's not terribly surprising, since deconstructing a list encoded using `Data` requires at least four builtin operations:
- `UnConstrData` to get the `Constr` args
- `Fst` to get the tag
- `Snd` to get the args
- `Head` to get the first arg
- ... additional `Head`/`Tail`s to get further args

I do think my benchmark is slightly worse for data lists than other usage, since it is working with a list of integers, rather than `Data` objects, so we have to do another builtin operation to extract the final integer. But I think the size of the difference is large enough that this is unlikely to be decisive.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Review by kwxm — CHANGES_REQUESTED** [review]
I'm not sure what's up with the golden files, but it should probably be fixed before merging.  Apart from that, it all looks OK.

**Comment by kwxm** [comment]
There's an opportunity for an argument about terminology here. Usually when I hear the word "benchmark" I think of something like Criterion benchmarks which run a program and see how much real-world time it takes.  Quite a few of the things we call benchmarks measure the cost of a PLC script, which is different (although we hope not too different), and we don't usually put them under `benchmark` in the cabal files.   Maybe we should call them something else, although I don't know what.

**Comment by michaelpj** [comment]
I have thought about that a bit. I do think they're benchmarks, they just happen to be _deterministic_ benchmarks. And for deterministic benchmarks you can treat them like golden tests, which has the advantage that you get a very clear signal of when they change.

File: plutus-benchmark/lists/test/Sum/Compiled left fold (Scott lists).budget.golden [hunk 1]
```diff
@@ -0,0 +1,2 @@
+({cpu: 150781800
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-benchmark/lists/test/Sum/Compiled left fold (Scott lists).budget.golden:1]: I think `cabal run`-ing a test executable does indeed not quite do what you want due to CWD issues.

I'll get rid of the special characters!

**Pull request merged**